RFC0003-sparse-roadmap/GPU.md

```diff
@@ -0,0 +1,147 @@
+are implemented using
+[hand-implemented](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu#L343) kernels.
+
+## PyTorch CUDA sparse data structures
+
+Both the CUDA and CPU variants of COO constructors utilize the same constructor
+(`sparse_coo_tensor` in SparseTensor.cpp). The main data of the COO tensor
+is stored in the `indices` and `values` objects.
+An example can be seen in
+[sparseElementwiseKernelScalar()](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/sparse/cuda/SparseCUDAApplyUtils.cuh#L76)
+function where the `TensorInfo` objects representing `indices` and `values`
+are passed into the CUDA kernel. Thus it can be seen that most of the functions are
+heavily tailored for the COO format and new functions must be written for new
+formats.
+
+When not working at the fine-grained level of the CUDA kernel, an `indices` tensor
+of type `LongTensor` and a `values` tensor of type `Tensor` can be accessed for
+accessing the values and indices of COO tensors respectively. These can then
+be directly interfaced with most libraries utilizing the COO tensor format
+since these libraries directly accept data pointers and their length to
+work with such data.
+
+Since the sparse matrix multiplication has been shown to be most optimal on a
+CSR format sparse matrix, PyTorch converts COO tensors into CSR tensors before
+calling such routines.
+
+## Usage of cuSPARSE in pytorch
+
+The only way the CSR format differs from the COO is its storage of the row indicies. 
+Since pytorch stores sparse matrices in COO by default, the pytorch implementation
+of sparse matrix-matrix multiplication
+[converts the row indices of the COO tensor](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu#L59)
+into CSR and
+[then performs the multiplication](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu#L96).
+Although the cuSPARSE interface
+can perform SpMM for COO tensors by specifying the appropriate macro denoting the layout,
+pytorch still performs conversion to CSR before calling the SpMM function, probably for
+performance reasons. Most of the cuSPARSE using functions can be found in
+[SparseCUDATensorMath.cu](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu)
+and [SparseCUDABlas.cu](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/sparse/cuda/SparseCUDABlas.cu).
+
+# Sparse tensor support for GPUs
+
+In this section we will have a glimpse at 3rd party libraries that support sparse
+matrix or sparse tensor operations. We refer to a 'sparse matrix' as a
+rank-2 tensor and 'sparse tensor' as a rank-n tensor. Sparse matrix and tensor
+implementations are being actively developed in libraries
+such as cuSPARSE, CUSP, MAGMA, Gingko and [ParTI!](https://github.com/hpcgarage/ParTI).
```

Comment:
Make the mentioned library names to links to the corresponding projects, as `ParTI`.